src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -80,9 +87,185 @@ private bool CanRewriteForEachAsFor(SyntaxNode forEachSyntax, TypeSymbol nodeExp
+
+            // !success
+            BoundExpression notSuccess = MakeUnaryOperator(UnaryOperatorKind.BoolLogicalNegation, forEachSyntax, method: null, boundSuccessVar, boundSuccessVar.Type);
+
+            // if (!success) goto outer_loop_continue;
+            BoundStatement checkAndBreak = RewriteIfStatement(forEachSyntax, notSuccess, continueOuterLoop, rewrittenAlternativeOpt: null, hasErrors: false).MakeCompilerGenerated();
+
+            //     V v = (V)(T)e.TryGetNext(out bool success);
+            //     if (!success) goto outer_loop_continue;
+            //     /* loop.Body */
+            //     /* loop.ContinueLabel: */
+            var innerLoopBodyBlock = CreateBlockDeclaringIterationVariables(iterationVariables, iterationVarDecl, checkAndBreak, rewrittenBody, loop.ContinueLabel, forEachSyntax);
+
+            // while (true)
+            // {
+            //     V v = (V)(T)e.TryGetNext(out bool success);
+            //     if (!success) goto outer_loop_continue;
+            //     /* loop.Body */
+            //     /* loop.ContinueLabel: */
+            // }
+            BoundStatement innerLoop = MakeWhileTrueLoop(loop, innerLoopBodyBlock);
+
+            // while (await e.WaitForNextAsync())
+            // {
+            //     /* inner loop */
+            //     outer_loop_continue:
+            // }
+            // /* loop.BreakLabel: */
+            BoundStatement outerLoopWithBreakLabel = RewriteWhileStatement(
+                loop: loop,
+                rewrittenCondition: awaitWaitForNext,
+                rewrittenBody: innerLoop,
+                breakLabel: loop.BreakLabel,
+                continueLabel: outerLoopContinueLabel,
+                hasErrors: false);
+
+            // try
+            // {
+            //     /* outer loop */
+            //     /* loop.BreakLabel: */
+            // }
+            // finally { await e.DisposeAsync(); }
+            // - OR variant without disposal: -
+            // /* outer loop */
+            // /* loop.BreakLabel: */
+            BoundStatement outerLoopOrTryFinally = outerLoopWithBreakLabel;
+            if (loop.EnumeratorInfoOpt.NeedsDisposeMethod && TryGetWellKnownTypeMember(syntax: null, WellKnownMember.System_IAsyncDisposable__DisposeAsync,
+                out MethodSymbol disposeAsyncMethod, location: forEachSyntax.AwaitKeyword.GetLocation()))
+            {
+                outerLoopOrTryFinally = WrapWithTryFinallyDispose(forEachSyntax, loop.EnumeratorInfoOpt, enumeratorType, boundEnumeratorVar, outerLoopWithBreakLabel, disposeAsyncMethod, loop.EnumeratorInfoOpt.DisposeAwaitableInfo);
+            }
+
+            // E e = ((C)(x)).GetAsyncEnumerator()
+            // try
+            // {
+            //     /* outer loop */
+            //     /* loop.BreakLabel: */
+            // }
+            // finally { await e.DisposeAsync(); }
+            BoundStatement result = new BoundBlock(
```

Comment:
I think `successVar` should be scoped to the innerLoopBodyBlock. There is not a lot of difference,, but it feels cleaner to put it on the narrowest containing block.